### PR TITLE
Dig post request

### PR DESCRIPTION
**Pull request by nanoNeel**
- Adds better way to display results for DIG
- Adds POST request to SUGAR
- Adds JSONJ support
- REFS: [POST request](https://github.com/nanoPayinc/NANOPAY/issues/4041), [JSONJ](https://github.com/nanoPayinc/NANOPAY/issues/4861)

### CODE REVIEW

File: src/foam/nanos/dig/DigWebAgent.java [hunk 1]
```diff
@@ -450,6 +510,17 @@ protected void outputException(X x, HttpServletResponse resp, Format format, Pri
       outputterHtml.outputEndTable();
       outputterHtml.outputEndHtml();
       out.println(outputterHtml.toString());
+    }else if ( format == Format.JSONJ ) {
```
Comment (jlhughes) [inline src/foam/nanos/dig/DigWebAgent.java]: space between } and 'else'

File: src/foam/nanos/dig/DigWebAgent.java [hunk 2]
```diff
@@ -213,6 +213,66 @@ public void execute(X x) {
...
+          String dataJsonJ[] = data.split("\\s*\\r?\\n");
+          for (String i:dataJsonJ){
+            if (i.startsWith("p(") && i.endsWith(")")) {
+              dataJson += i.substring(2, i.length()-1) + ',';
+            }
+          }
+          dataJson += "]";
+         
+        // JSON part from above
+        JSONParser jsonParser = new JSONParser();
```
Comment (jlhughes) [inline src/foam/nanos/dig/DigWebAgent.java]: indent

File: src/foam/nanos/dig/DigWebAgent.java (hunk 1 above)
Comment (jlhughes) [reply src/foam/nanos/dig/DigWebAgent.java]: hard to tell from this PR, what is the difference between this JSONJ block and that above?

File: src/foam/nanos/dig/SUGAR.js [hunk 3]
```diff
@@ -231,27 +255,70 @@ foam.CLASS({
...
+      hidden: true,
+      label: 'Send POST Request',
+      code: async function() {
+        if ( ! (this.postURL === '') ) {
+          var req = this.HTTPRequest.create({
+            url: this.appConfig.URL.value + this.postURL.substring(1),
+            method: 'POST',
+            contentType: 'url',
+            payload: this.postData.substring(1),
+          }).send();
```
Comment (jlhughes) [inline src/foam/nanos/dig/SUGAR.js:304]: Doesn't have to happen in this PR (can be future ticket), but we need to add http authentication info with the sessionid so these posts can proceed with an additional login.

File: src/foam/nanos/dig/DigWebAgent.java [hunk 4]
```diff
@@ -213,6 +213,66 @@ public void execute(X x) {
           outputException(x, resp, format, out, error);
 
           return;
+        } else if (Format.JSONJ == format ) {
+          String dataJson = "[";
+          String dataJsonJ[] = data.split("\\s*\\r?\\n");
+          for (String i:dataJsonJ){
+            if (i.startsWith("p(") && i.endsWith(")")) {
```
Comment (jlhughes) [inline src/foam/nanos/dig/DigWebAgent.java]: Also, you shouldn't need to do this. Have a look in the 'replay' of the JDAO as it has logic to apply the json/j operations, unless of course it does the same thing.

File: src/foam/nanos/dig/DigWebAgent.java (hunk 1 above)
Comment (nanoNeel) [reply src/foam/nanos/dig/DigWebAgent.java]: This JSONJ block is for the outputException method which is called when ever there is an error. Without this block the webpage won't load and you would have no idea if it failed or not. The JSONJ block above is for the put command.

File: src/foam/nanos/dig/DigWebAgent.java [hunk 5]
```diff
@@ -213,6 +213,65 @@ public void execute(X x) {
           outputException(x, resp, format, out, error);
 
           return;
+        } else if (Format.JSONJ == format ) {
+          String dataJson = "[";
+          String dataJsonJ[] = data.split("\\r?\\n");
+          for (String i:dataJsonJ){
+            i = i.trim();
+            dataJson += i.substring(2, i.length()-1) + ',';
```
Comment (jlhughes) [inline src/foam/nanos/dig/DigWebAgent.java]: Ok, i preferred the test for 'p'
`         if (i.startsWith("p(") && i.endsWith(")")) {`
as that ensures we only put 'p' operations, the blind substring can let through 'r' operations on command 'put'.  
Is this really what the JDAO does?

**Review by jlhughes — APPROVED** [review]

**Pull request merged**